feat(curriculum): Extend solution explanation for Convert HTML Entities challenge (#35835)
freeCodeCamp/freeCodeCamp: guide/english/certifications/javascript-algorithms-and-data-structures/intermediate-algorithm-scripting/convert-html-entities/index.md

## ![:beginner:](https://forum.freecodecamp.com/images/emoji/emoji_one/beginner.png?v=3 ":beginner:") Basic Code Solution:
```javascript
function convertHTML(str) {
  // Split by character to avoid problems.

  var temp = str.split('');

>   // Since we are only checking for a few HTML elements I used a switch

Reviewer: ```suggestion
  // Since we are only checking for a few HTML elements, use a switch
```

## ![:sunflower:](https://forum.freecodecamp.com/images/emoji/emoji_one/sunflower.png?v=3 ":sunflower:") Intermediate Code Solution:

```javascript
function convertHTML(str) {
  // Use Object Lookup to declare as many HTML entities as needed.
  const htmlEntities = {
    "&" : '&amp;',
    "<" : "&lt;",
    ">" : "&gt;",
    '"' : "&quot;",
    "'" :"&apos;"
  }
  // Using a regex, replace characters with it's corresponding html entity
  return str.replace(new RegExp("([&<>\"'])", "g"), match => htmlEntities[match]);
}

// test here
convertHTML("Dolce & Gabbana");
```

> ![:rocket:](https://forum.freecodecamp.com/images/emoji/emoji_one/rocket.png?v=3 ":rocket:") <a href='https://repl.it/repls/WoozyDelightfulSoftwareengineer' target='_blank' rel='nofollow'>Run Code</a>

Reviewer: ```suggestion
```
We're removing external links from now on, so this line should be deleted

*   The second arguments for `replace()` is a function with the matched character as a parameter. It returns the correspondant entity from `htmlEntities`.

#### Relevant Links

*   <a href='https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace' target='_blank' rel='nofollow'>str.replace()</a>
*   <a href='https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp' target='_blank' rel='nofollow'>Regular Expressions</a>

## ![:rotating_light:](https://forum.freecodecamp.com/images/emoji/emoji_one/rotating_light.png?v=3 ":rotating_light:") Advanced Code Solution:

```javascript
function convertHTML(str) {
  // Use Object Lookup to declare as many HTML entities as needed.
  const htmlEntities={
    '&':'&amp;',
    '<':'&lt;',
    '>':'&gt;',
    '"':'&quot;',
    '\'':"&apos;"
  };
  //Use map function to return a filtered str with all entities changed automatically.
  return str.split('').map(entity => htmlEntities[entity] || entity).join('');
}

// test here
convertHTML("Dolce & Gabbana");
```

> ![:rocket:](https://forum.freecodecamp.com/images/emoji/emoji_one/rocket.png?v=3 ":rocket:") <a href='https://repl.it/CLnR/0' target='_blank' rel='nofollow'>Run Code</a>

Reviewer: ```suggestion
```
Same for this
